File: python/find_first_palindromic_string_in_the_array.py
class Solution:
    def firstPalindrome(self, words: list[str]) -> str:
        for word in words:
            if self.is_parli(word):
                return word
        return ""

    def is_parli(self, word: str) -> bool:
        i: int = 0
        j: int = len(word) - 1
        while i < j:
            if word[i] != word[j]:
                return False
            i += 1
            j -= 1
        return True


def test_firstPalindrome_case_1():
    # arrange
    words: list[str] = ["abc", "car", "ada", "racecar", "cool"]
    expected: str = "ada"

    # act
    solution = Solution()
    actual = solution.firstPalindrome(words)

    # assert
    assert expected == actual


def test_firstPalindrome_case_2():
    # arrange
    words: list[str] = ["notapalindrome", "racecar"]
    expected: str = "racecar"

    # act
    solution = Solution()
    actual = solution.firstPalindrome(words)

    # assert
    assert expected == actual


def test_firstPalindrome_case_3():
    # arrange
    words: list[str] = ["def", "ghi"]
    expected: str = ""

    # act
    solution = Solution()
    actual = solution.firstPalindrome(words)

    # assert
    assert expected == actual

File: rust/src/code/find_first_palindromic_string_in_the_array.rs
struct Solution;

impl Solution {
    pub fn first_palindrome(words: Vec<String>) -> String {
        for word in words {
            let word: Vec<char> = word.chars().collect();
            if Self::is_pali(&word) {
                return word.iter().collect::<String>();
            }
        }
        "".to_string()
    }

    fn is_pali(word: &Vec<char>) -> bool {
        if word.len() == 0 {
            return true;
        }
        let mut i: usize = 0;
        let mut j: usize = word.len() - 1;
        while i < j {
            if word[i] != word[j] {
                return false;
            }
            i += 1;
            j -= 1;
        }
        true
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn test_first_palindrome_case_1() {
        // arrange
        let words: Vec<String> = vec![
            "abc".to_string(),
            "car".to_string(),
            "ada".to_string(),
            "racecar".to_string(),
            "cool".to_string(),
        ];
        let expected: String = "ada".to_string();

        // act
        let actual = Solution::first_palindrome(words);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_first_palindrome_case_2() {
        // arrange
        let words: Vec<String> = vec!["notapalindrome".to_string(), "racecar".to_string()];
        let expected: String = "racecar".to_string();

        // act
        let actual = Solution::first_palindrome(words);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_first_palindrome_case_3() {
        // arrange
        let words: Vec<String> = vec!["def".to_string(), "ghi".to_string()];
        let expected: String = "".to_string();

        // act
        let actual = Solution::first_palindrome(words);

        // assert
        assert_eq!(expected, actual);
    }
}
